external/fv3fit/tests/test_bptt.py

```diff
@@ -0,0 +1,272 @@
+import pytest
+import xarray as xr
+import tensorflow as tf
+import numpy as np
+from datetime import timedelta
+from cftime import DatetimeProlepticGregorian as datetime
+from fv3fit.keras._models.recurrent import BPTTModel
+import tempfile
+import fv3fit
+
+
+@pytest.fixture
+def sample_dim_name():
+    return "sample"
+
+
+@pytest.fixture
+def dt():
+    return timedelta(seconds=1)
+
+
+@pytest.fixture
+def train_dataset(sample_dim_name, dt):
+    n_samples = 6
+    n_times = 12
+    n_z = 16
+    t_start = datetime(2020, 1, 1)
+    ds = xr.Dataset(
+        data_vars={
+            "air_temperature": xr.DataArray(
+                np.random.randn(n_samples, n_times, n_z),
+                dims=[sample_dim_name, "time", "z"],
+                attrs={"units": "K"},
+            ),
+            "specific_humidity": xr.DataArray(
+                np.random.randn(n_samples, n_times, n_z),
+                dims=[sample_dim_name, "time", "z"],
+                attrs={"units": "kg/kg"},
+            ),
+            "air_temperature_tendency_due_to_model": xr.DataArray(
+                np.random.randn(n_samples, n_times, n_z),
+                dims=[sample_dim_name, "time", "z"],
+            ),
+            "specific_humidity_tendency_due_to_model": xr.DataArray(
+                np.random.randn(n_samples, n_times, n_z),
+                dims=[sample_dim_name, "time", "z"],
+            ),
+            "air_temperature_tendency_due_to_nudging": xr.DataArray(
+                np.random.randn(n_samples, n_times, n_z),
+                dims=[sample_dim_name, "time", "z"],
+            ),
+            "specific_humidity_tendency_due_to_nudging": xr.DataArray(
+                np.random.randn(n_samples, n_times, n_z),
+                dims=[sample_dim_name, "time", "z"],
+            ),
+            "a": xr.DataArray(
+                np.random.randn(n_samples, n_times), dims=[sample_dim_name, "time"],
+            ),
+            "b": xr.DataArray(
+                np.random.randn(n_samples, n_times), dims=[sample_dim_name, "time"],
+            ),
+        },
+        coords={"time": np.asarray([t_start + i * dt for i in range(n_times)])},
+    )
+    return ds
+
+
+def test_predict_model_gives_tendency_of_train_model(
+    train_dataset, sample_dim_name, dt
```

Comment:
The pattern of using fixtures to pass "metadata" like dt and sample_dim_name of a constructed test object "train_dataset" tightly couples all these fixtures. This makes it hard to modify one fixture without changing the others. Checkout the git log of the other places in the code that use this pattern (e.g. fv3fit/tests/training/conftest.py). You will note that most commits touching such code require changing multiple fixtures at once, a sign of excessive coupling. It is one reason why the root level tests were hard to maintain.

Remember how I used to be against fixtures? It was mostly because of this pattern. However, fixtures are not the problem. Rather, the problem is that this test function assumes that train_dataset was constructed from the other two arguments. It takes 3 arguments, but only really has 2 independent pieces of data: "sample_dim" and "dt". This constraint makes it brittle.

This approach is more flexible, and removes any implied coupling between `train_dataset`, `sample_dim_name`, and `dt`:
```
 test_predict_model_gives_tendency_of_train_model(
    sample_dim_name, dt
):
    train_dataset = get_train_dataset(sample_dim_name, dt)
```

Reply:
The problem gets worse when more tests use these fixtures. Each tests enforces a slightly different constraint between the fixtures, and ultimately it can become impossible to make any changes to the fixture hierarchy without breaking some test.

Reply:
Yeah, I like this change. Will do.